arch/xtensa/src/esp32s3/esp32s3_ble_adapter.c

```diff
@@ -0,0 +1,2710 @@
+ * Name: esp32s3_bt_controller_deinit
+ *
+ * Description:
+ *   Deinit BT controller.
+ *
+ * Input Parameters:
+ *   None
+ *
+ * Returned Value:
+ *   Zero (OK) is returned on success. Otherwise, -1 (ERROR) is returned.
+ *
+ ****************************************************************************/
+
+int esp32s3_bt_controller_deinit(void)
+{
+  if (btdm_controller_status != ESP_BT_CONTROLLER_STATUS_INITED)
+    {
+      return ERROR;
+    }
+
+  btdm_controller_deinit();
+
+  esp32s3_bt_controller_deinit_internal();
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: esp32s3_bt_controller_deinit_internal
+ *
+ * Description:
+ *   Deinit BT internal controller.
+ *
+ * Input Parameters:
+ *    None
+ *
+ * Returned Value:
+ *    None
+ *
+ ****************************************************************************/
+
+static void esp32s3_bt_controller_deinit_internal(void)
+{
+  esp32s3_periph_module_disable(PERIPH_BT_MODULE);
+
+  if (g_lp_stat.phy_enabled)
+    {
+      esp32s3_phy_disable();
+      g_lp_stat.phy_enabled = 0;
+    }
+
+  /* deinit low power control resources */
+
+  do
+    {
+#if CONFIG_MAC_BB_PD
+      if (g_lp_cntl.mac_bb_pd)
+        {
+          btdm_deep_sleep_mem_deinit();
+          g_lp_cntl.mac_bb_pd = 0;
+        }
+#endif
+
+#ifdef CONFIG_PM
+      if (g_lp_cntl.no_light_sleep)
+        {
+          if (g_light_sleep_pm_lock != NULL)
+            {
+              esp_pm_lock_delete(g_light_sleep_pm_lock);
+              g_light_sleep_pm_lock = NULL;
+            }
+        }
+
+      if (g_pm_lock != NULL)
+        {
+          esp_pm_lock_delete(g_pm_lock);
+          g_pm_lock = NULL;
+          g_lp_stat.pm_lock_released = 0;
+        }
+#endif
+
+      if (g_lp_cntl.wakeup_timer_required)
+        {
+          if (g_lp_stat.wakeup_timer_started)
+            {
+              esp_timer_stop(g_btdm_slp_tmr);
+            }
+
+          g_lp_stat.wakeup_timer_started = 0;
+          esp_timer_delete(g_btdm_slp_tmr);
+          g_btdm_slp_tmr = NULL;
+        }
+
+      if (g_lp_cntl.enable)
+        {
+          btdm_vnd_offload_task_deregister(BTDM_VND_OL_SIG_WAKEUP_TMR);
+          if (g_wakeup_req_sem != NULL)
+            {
+              semphr_delete_wrapper(g_wakeup_req_sem);
+              g_wakeup_req_sem = NULL;
+            }
+        }
+
+      if (g_lp_cntl.lpclk_sel == BTDM_LPCLK_SEL_XTAL)
+        {
+#ifdef CONFIG_BT_CTRL_MAIN_XTAL_PU_DURING_LIGHT_SLEEP
+          if (g_lp_cntl.main_xtal_pu)
+            {
+              esp_sleep_pd_config(ESP_PD_DOMAIN_XTAL, ESP_PD_OPTION_OFF);
+              g_lp_cntl.main_xtal_pu = 0;
+            }
+#endif
+
+          btdm_lpclk_select_src(BTDM_LPCLK_SEL_RTC_SLOW);
+          btdm_lpclk_set_div(0);
+#if CONFIG_ESP32S3_WIFI_BT_COEXIST
+          coex_update_lpclk_interval();
+#endif
+        }
+
+      g_btdm_lpcycle_us = 0;
+    }
+  while (0);
```

Comment:
why do we need `do {} while (0)` wrapper if no `break` is used?

Reply:
We don't need... It came from the bt adapter of ESP-IDF. I'll remove it.